### DISCUSSION TITLE
"strict": true collides with mdast types

### DISCUSSION THREAD

**Discussion by 205g0** [syntax-tree/unist#52] [Q&A] [answered]
This is related to the prior discussion. Please check the example repo https://github.com/205g0/mdast-example and there `index.ts`:

You should see in your editor following contextual error in `index.ts` at line 21 over `(node: Heading)` which says:

_[tsserver 2345] [E] Argument of type '(node: Heading) => void' is not assignable to parameter of type '(value: Node<Data>, index: number, array: Node<Data>[]) => void'.   Types of parameters 'node' and 'value' are incompatible. Type 'Node<Data>' is missing the following properties from type 'Heading': depth, children_

If you remove `"strict": true`, this error goes away. __Be aware that I already broke down the `strict` compiler option in all its single strict options at the top of `tsconfig.json`__ in order to isolate the causing option, which is:

`"strictFunctionTypes": true`

Remove this and the error goes away. Not sure if this is intended and I misunderstood something.

**Comment by wooorm** [comment]
But `selectAll('root > *', ast)` could result in other nodes as well, though, right? Not just headings?

**Comment by 205g0** [comment, marked as answer]
Yes, you are right, I just realized once I hit submit. Then how do I properly type this `node`?

**Reply by wooorm** [reply]
See: https://unifiedjs.com/learn/guide/syntax-trees-typescript/, https://unifiedjs.com/learn/recipe/tree-traversal-typescript/, and https://unifiedjs.com/learn/recipe/narrow-node-typescript/.

**Reply by 205g0** [reply]
Yes, thanks, I checked them I few times and I can do this again, but the docs don't give an example on thes use case. So, please check my comment below. Anything I have tried (interface Data etc.) triggers an error over line 22 over `.depth` and I couldn't get it right.

**Reply by 205g0** [reply]
re the use case again: the docs do not say how do I properly narrow down an mdast type, e.g. in a forEach

**Reply by wooorm** [reply]
Well first of all you don’t need `select` or any type annotations:

```js
ast.children.forEach((node) => {
  if (node.type === 'heading' && node.depth == 2) {
    console.log('I found a heading with level 2')
  }
})
```

As the docs I linked to mention, you can use `unist-util-is`:

```ts
selectAll('root > *', ast).forEach((node) => {
  if (is<Heading>(node, 'heading') && node.depth === 2) {
    console.log('I found a heading with level 2')
  }
})
```

**Reply by 205g0** [reply]
Thanks for your help, I just got it from docs and pushed the solution to that repo. Sorry for having been to fast. Could I also do it without `is` somehow?

Re the `selectAll`, yes right, I need it in my full app, in this example it is not required.

**Reply by 205g0** [reply]
I ask since in TS you use usually `typeof` typeguards and I was just wondering if I could somehow do it myself. But to get the actual type I would need to read the prop `.type` but that's not enough to typeguard and/or sasert.

**Reply by wooorm** [reply]
```ts
import type {Content} from 'mdast'

const nodes = selectAll('root > *', ast) as Content[]
```

Cast it, something like this probably works

**Reply by 205g0** [reply]
ok thanks till now! I play around a bit and may come back later, again thanks for your support, highly appreciated!

**Reply by wooorm** [reply]
`unist-util-visit` is much smarter btw: ``visit(mdast, 'heading', (node) => { /* TS knows `node` is a heading */ })``

**Reply by 205g0** [reply]
Ok, I refactored with `visit` only and this is much better. I could handle my transformation logic just with a `switch` statement checking if, e.g., `case node.type = 'heading'`, etc. This works very well with all node types. But once I introduced `remark-frontmatter` to the AST and checking for the frontmatter I cannot read the value from the node:

```
...
case node.type == 'yaml':
  console.log(node.value)
...
```
I get this error on `node.value`: _[tsserver 2339] [E] Property 'value' does not exist on type 'Root | Paragraph | Heading | ThematicBreak | Blockquote | List | Table | HTML | Code | YAML | ... 16 more ... | FootnoteReference'. Property 'value' does not exist on type 'Root'._

This error is confusing since `YAML` is explicitly stated. I know that there is package to extract the frontmatter but I thought I can just read the value and do it myself and also because this node type is not nested.

Also, when I narrow with `is`...
```
import { YAML } from 'mdast'
...
case is<YAML>(node, 'yaml'):
  console.log(node.value)
```
...I get the same error. There're also no conflicts with modules/types in the `YAML` namespace.

**Reply by 205g0** [reply]
When I do another extra assertion like...
```
case is<YAML>(node, 'yaml'):
  const yaml = node as YAML
  console.log(yaml)
```
The editor errors goes away, so it works at CT but not at RT where I get _Cannot handle unknown node `yaml`_

**Reply by 205g0** [reply]
Ok, I think I got my problem: `remark-frontmatter` does not expect an AST but actually a __markdown__ string...

**Reply by 205g0** [reply]
No, I was the reply before wrong, `remark-frontmatter` expects an AST and creates a perfectly fine AST node:

```
{
  type: 'root',
  children: [
    {
      type: 'yaml',
      value: 'name: Max',
      position: [Object]
    },
    ...
```
This is really weird...

**Reply by 205g0** [reply]
FWIW, I get `Error: Cannot handle unknown node 'yaml'` also without any `console.log`, so just with an empty `case`. I go to bed now and try again tmr...

**Reply by 205g0** [reply]
Now, I refactored again, kicked-out any yaml, frontmatter, frontmatter processing and now I get the same error as above on `.depth` at:
```
switch (true) {
  case is<Heading>(node, 'heading'):
    if (node.depth == 1) {
      ...
```
The error again: _[tsserver 2339] [E] Property 'depth' does not exist on type 'Root | Paragraph | Heading | ThematicBreak | Blockquote | List | Table | HTML | Code | YAML | ... 16 more ... | FootnoteReference'. Property 'depth' does not exist on type 'Root'._

🤯

Edit: I use `visit` to walk the AST

**Reply by ChristianMurphy** [reply]
> ```ts
> switch (true) {
>   case is<Heading>(node, 'heading'):
>     if (node.depth == 1) {
>       ...
> ```


TypeScript doesn't support narrowing types in `case` the way you show.
See this playground for an example, note the value of `something` in the different blocks https://www.typescriptlang.org/play?#code/PTAEBcAsEsGdQA4CcCmATaBjAhuFocA7UQ7JJAewHdRtiBXQga0OuPAE8F9wLbRY4JNEIBzAFAAzRpnDQKxOAGUhI0QAoAbtgA29FAC5QjFmwCUR7XvxwBqsaADeoVOHpJ2XFBUmgr+0ABeYNAAckFhMVCAX3FxEFAqfCo6cAg+UnJqCBhYcUwFQQEKAFsUKDUjE1YqYkCSeh0dOOhfdWV7DVhS8pgxMzMncVAR0ASK+FtMyiS0YdHusoqxcVjxWCpocExIUHUhfUHHeZGcWBtYFUiunuXRCyHRp7GwCdApinAT58XetW-YrEgA

You need to use an `if` to get a predicate to narrow the value in scope.

> I use visit to walk the AST

A good idea, `visit` can handle narrowing types itself, which makes using specific node types easier

**Reply by wooorm** [reply]
@205g0 your javascript is wrong, switch doesn't work like that. https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/switch.

**Reply by 205g0** [reply]
@wooorm right,`switch`/ `case` are the way around and weird anyway. @ChristianMurphy thanks Christian, I refactored again and I think the golden rule is just to use `if`s and `visit` and call it a day. While you can do anything with `visit` I like the stuff `selectAll` offers but I could create some clean code just with `visit` as well, btw how do you get syntax highlighting in your code fields?

**Reply by wooorm** [reply]
This example shows both how switch in JavaScript works, and how to attach a code language to code:

````markdown
```js
// Assuming `node` is a known mdast node such as `mdast.Content`
switch (node.type) {
  case 'heading':
    // TS should know this is a heading.
    break
  case 'yaml':
    // TS should know this is YAML.
    break
}
```
````
